faidx.c

```diff
@@ -175,16 +263,49 @@ static int write_output(faidx_t *faid, output *out, const char *name, const int
     if (rev && seq_len > 0) {
         reverse_complement(seq, seq_len);
     }
+
+    if (out->gopt->write_index && !allocidx(out->idxdata)) {    //index data allocation
+        fprintf(stderr, "[faidx] Failed to allocate memory for indexing.\n");
+        goto end;
+    }
+
     //write the name
     len = ksprintf(&out->buffer, "%c%s%s\n", format == FAI_FASTA ? '>' : '@', name, rev ? neg_strand_name : pos_strand_name);
-    if (wrappedwrite(out, out->buffer.s, out->buffer.l) < len) {
-        fprintf(stderr,"[faidx] Failed to write buffer\n");
-        goto exit;
+    if (bgzf_write(out->bgzf_fp, out->buffer.s, out->buffer.l) < len) {
+        fprintf(stderr,"[faidx] Failed to write buffer.\n");
+        goto end;
+    }
+
+    if (out->gopt->write_index) {                                   //index data for current sequence
+        if (!(tmp = malloc(len - 1))) {
+            fprintf(stderr, "[faidx] Failed to allocate memory.\n");
+            goto end;
+        }
+        strncpy(tmp, out->buffer.s + 1, len - 2);                   //copy name without start marker
+        tmp[len - 2] = '\0';
+        ks_clear(&out->buffer);
+
+        out->idxdata->indx[out->idxdata->n].name = tmp;
+        out->idxdata->indx[out->idxdata->n].seq_offset += len;      //last offset + length of name line
+        out->idxdata->indx[out->idxdata->n].seq_length = seq_len;
     }
+
     ks_clear(&out->buffer);
     //write bases
     if ((ret = write_line(faid, out, seq, name, ignore, wrap_len, seq_len) == EXIT_FAILURE)) {
-        goto exit;
+        goto end;
+    }
+
+    if (out->gopt->write_index) {
+        out->idxdata->indx[out->idxdata->n].line_length = seq_len < wrap_len ? seq_len : wrap_len;
+        if (out->idxdata->format != FAI_FASTA) {                    //fastq, sequence offset + seq len and '\n's
+            out->idxdata->indx[out->idxdata->n].qual_offset = out->idxdata->indx[out->idxdata->n].seq_offset +
+             seq_len + seq_len / wrap_len + ((seq_len % wrap_len) ? 1 : 0);
+        } else {                                                    //fasta, next sequence's temp offset
+            out->idxdata->indx[out->idxdata->n + 1].seq_offset = out->idxdata->indx[out->idxdata->n].seq_offset +
+             seq_len + seq_len / wrap_len + ((seq_len % wrap_len) ? 1 : 0);
+            ++out->idxdata->n;
```

Comment:
Why are there two places incrementing `out->idxdata->n`?  It's confusing as fasta does it in one place and fastq in another, leading to code duplication.  Ultimately though we only have one index entry regardless of format.

Again, if we had a single function to allocate and return the next index entry (allocidx()) then this could increment n itself.